cpp/src/structure/sort_and_remove_multi_edges_impl.cuh

```diff
@@ -0,0 +1,317 @@
+{
+  rmm::device_uvector<size_t> hash_counts(num_hash_buckets, stream_view);
+  thrust::fill(rmm::exec_policy(stream_view), hash_counts.begin(), hash_counts.end(), size_t{0});
+
+  auto hash_counts_view = raft::device_span<size_t>(hash_counts.data(), hash_counts.size());
+  thrust::for_each(rmm::exec_policy(stream_view),
+                   iter_first,
+                   iter_last,
+                   [num_hash_buckets, hash_op, hash_counts_view] __device__(auto value) {
+                     atomicAdd(&hash_counts_view[hash_op(value) % num_hash_buckets], size_t{1});
+                   });
+
+  thrust::exclusive_scan(
+    rmm::exec_policy(stream_view), hash_counts.begin(), hash_counts.end(), hash_counts.begin());
+
+  return hash_counts;
+}
+
+template <typename vertex_t>
+struct hash_src_dst_pair {
+  int32_t __device__ operator()(thrust::tuple<vertex_t, vertex_t> t) const
+  {
+    vertex_t pair[2];
+    pair[0] = thrust::get<0>(t);
+    pair[1] = thrust::get<1>(t);
+    cuco::detail::MurmurHash3_32<vertex_t*> hash_func{};
+    return hash_func.compute_hash(reinterpret_cast<std::byte*>(pair), 2 * sizeof(vertex_t));
+  }
+};
+
+template <typename vertex_t>
+std::tuple<rmm::device_uvector<vertex_t>, rmm::device_uvector<vertex_t>>
+sort_and_remove_multi_edges(raft::handle_t const& handle,
+                            rmm::device_uvector<vertex_t>&& edgelist_srcs,
+                            rmm::device_uvector<vertex_t>&& edgelist_dsts,
+                            size_t mem_frugal_threshold)
+{
+  auto pair_first = thrust::make_zip_iterator(edgelist_srcs.begin(), edgelist_dsts.begin());
+
+  if (edgelist_srcs.size() > mem_frugal_threshold) {
+    // Tuning parameter to address high frequency multi-edges
+    size_t num_hash_buckets{16};
+
+    auto hash_counts = compute_hash_sizes(pair_first,
+                                          pair_first + edgelist_srcs.size(),
+                                          num_hash_buckets,
+                                          hash_src_dst_pair<vertex_t>{},
+                                          handle.get_stream());
+
+    auto pivot =
+      static_cast<int32_t>(thrust::distance(hash_counts.begin(),
+                                            thrust::lower_bound(handle.get_thrust_policy(),
+                                                                hash_counts.begin(),
+                                                                hash_counts.end(),
+                                                                edgelist_srcs.size() / 2)));
+
+    auto second_first = detail::mem_frugal_partition(pair_first,
+                                                     pair_first + edgelist_srcs.size(),
+                                                     hash_src_dst_pair<vertex_t>{},
+                                                     pivot,
+                                                     handle.get_stream());
```

Comment:
You can use https://github.com/rapidsai/cugraph/blob/branch-23.12/cpp/include/cugraph/utilities/shuffle_comm.cuh#L782 with num_groups = 2. Then you just need to provide a `key_to_group_id` functor.

Reply:
My concern with setting num_groups = 2 was that we might get a pretty bad skew in the multi-edges.  Consider an rmat graph.  We're going to get certain edges that occur with great frequency, and other edges that occur with lesser frequency.  I thought that by first grouping by some larger number and then finding the median break point we'll be closer to a 50/50 split at the pivot.

I certainly could use `groupby_and_count` instead of recreating it here (my `compute_hash_sizes` function is essentially the same.

Reply:
You mean there can be millions (or billions) of edges between a unique (src, dst) pair? Yeah... this might be possible if there are two really big institutes and there are a huge number of transactions between the two institutes.

But I don't think this will happen in synthetic R-mat graphs (we are not partitioning based on just src or dst but src, dst pair, so high degree vertices won't have much impact on partitioning quality).

Yeah... and calling with num_groups = 2 vs some large number can be a compute vs solution robustness trade-offs. num_goups = 2 will be cheaper than than num_groups = 128 (2^1 vs 2^7). And in the theoretical worst case (i.e. all the edges are between the same src, dst pair), we will need 2x memory of the edge list vs 1.5x. And if there are very high multiplicity edge pairs, we may be sometimes able to combine those edges and create a weighted graph. Just a thought material. Not sure what's the right trade-off point.